### PR TITLE
Fix LSP on Windows

### LINKED ISSUE #3918: [VS Code Extension] Opening document containing graphql tagged literal causes 'Request canceled' errors
I'm really exited about the Relay VS Code extension, so I wanted to try it out early, but unfortunately ran into an issue.

### Setup

I setup a fresh project using Relay (relay-compiler@main), installed VS Code insiders and the extension. I've not modified any extension settings (besides `relay.outputLevel` for troubleshooting) and I can execute the `relay-compiler` just fine from the command line. I also do not have any conflicting extensions (the Relay extension is the only one) and my `relay-compiler` config is in the `package.json` file.

### Problem manifestation

Looking at the logs the extension seems to be starting correctly and the ` textDocument/didOpen` action seems to be successful, but then it starts throwing hundreds of "Request canceled" errors for the `textDocument/codeAction` action (I've shortened the log output below). The extension still accepts commands like restarting the language server, but the main features like go to definition, etc. do not appear to be working.

#### Log output

```shell
Starting the Relay GraphQL extension...
Searching for the relay-compiler starting at: c:\src\myapp\app
You have a pre-release version of the relay-compiler package installed. We are unable to confirm if this version is compatible with the Relay VSCode Extension. Proceeding on the assumption that you know what you are doing.
Using relay binary: c:\src\myapp\app\node_modules\relay-compiler\win-x64\relay.exe
Starting the Relay Langauge Server with these options: {"options":{"cwd":"c:\\src\\myapp\\app"},"command":"c:\\src\\myapp\\app\\node_modules\\relay-compiler\\win-x64\\relay.exe","args":["lsp","--output=debug"]}
[0m[36m[DEBUG] [0mInitialized stdio transport layer
[0m[36m[DEBUG] [0m< {"jsonrpc":"2.0","id":0,"method":"initialize","params":{"processId":17128,"clientInfo":{"name":"Visual Studio Code - Insiders","version":"1.68.0-insider"},"locale":"en-us","rootPath":"c:\\src\\myapp\\app","rootUri":"file:///c%3A/src/myapp/app","capabilities":{"workspace":{"applyEdit":true,"workspaceEdit":{"documentChanges":true,"resourceOperations":["create","rename","delete"],"failureHandling":"textOnlyTransactional","normalizesLineEndings":true,"changeAnnotationSupport":{"groupsOnLabel":true}},"didChangeConfiguration":{"dynamicRegistration":true},"didChangeWatchedFiles":{"dynamicRegistration":true},"symbol":{"dynamicRegistration":true,"symbolKind":{"valueSet":[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26]},"tagSupport":{"valueSet":[1]}},"codeLens":{"refreshSupport":true},"executeCommand":{"dynamicRegistration":true},"configuration":true,"workspaceFolders":true,"semanticTokens":{"refreshSupport":true},"fileOperations":{"dynamicRegistration":true,"didCreate":true,"didRename":true,"didDelete":true,"willCreate":true,"willRename":true,"willDelete":true}},"textDocument":{"publishDiagnostics":{"relatedInformation":true,"versionSupport":false,"tagSupport":{"valueSet":[1,2]},"codeDescriptionSupport":true,"dataSupport":true},"synchronization":{"dynamicRegistration":true,"willSave":true,"willSaveWaitUntil":true,"didSave":true},"completion":{"dynamicRegistration":true,"contextSupport":true,"completionItem":{"snippetSupport":true,"commitCharactersSupport":true,"documentationFormat":["markdown","plaintext"],"deprecatedSupport":true,"preselectSupport":true,"tagSupport":{"valueSet":[1]},"insertReplaceSupport":true,"resolveSupport":{"properties":["documentation","detail","additionalTextEdits"]},"insertTextModeSupport":{"valueSet":[1,2]}},"completionItemKind":{"valueSet":[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25]}},"hover":{"dynamicRegistration":true,"contentFormat":["markdown","plaintext"]},"signatureHelp":{"dynamicRegistration":true,"signatureInformation":{"documentationFormat":["markdown","plaintext"],"parameterInformation":{"labelOffsetSupport":true},"activeParameterSupport":true},"contextSupport":true},"definition":{"dynamicRegistration":true,"linkSupport":true},"references":{"dynamicRegistration":true},"documentHighlight":{"dynamicRegistration":true},"documentSymbol":{"dynamicRegistration":true,"symbolKind":{"valueSet":[1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26]},"hierarchicalDocumentSymbolSupport":true,"tagSupport":{"valueSet":[1]},"labelSupport":true},"codeAction":{"dynamicRegistration":true,"isPreferredSupport":true,"disabledSupport":true,"dataSupport":true,"resolveSupport":{"properties":["edit"]},"codeActionLiteralSupport":{"codeActionKind":{"valueSet":["","quickfix","refactor","refactor.extract","refactor.inline","refactor.rewrite","source","source.organizeImports"]}},"honorsChangeAnnotations":false},"codeLens":{"dynamicRegistration":true},"formatting":{"dynamicRegistration":true},"rangeFormatting":{"dynamicRegistration":true},"onTypeFormatting":{"dynamicRegistration":true},"rename":{"dynamicRegistration":true,"prepareSupport":true,"prepareSupportDefaultBehavior":1,"honorsChangeAnnotations":true},"documentLink":{"dynamicRegistration":true,"tooltipSupport":true},"typeDefinition":{"dynamicRegistration":true,"linkSupport":true},"implementation":{"dynamicRegistration":true,"linkSupport":true},"colorProvider":{"dynamicRegistration":true},"foldingRange":{"dynamicRegistration":true,"rangeLimit":5000,"lineFoldingOnly":true},"declaration":{"dynamicRegistration":true,"linkSupport":true},"selectionRange":{"dynamicRegistration":true},"callHierarchy":{"dynamicRegistration":true},"semanticTokens":{"dynamicRegistration":true,"tokenTypes":["namespace","type","class","enum","interface","struct","typeParameter","parameter","variable","property","enumMember","event","function","method","macro","keyword","modifier","comment","string","number","regexp","operator"],"tokenModifiers":["declaration","definition","readonly","static","deprecated","abstract","async","modification","documentation","defaultLibrary"],"formats":["relative"],"requests":{"range":true,"full":{"delta":true}},"multilineTokenSupport":false,"overlappingTokenSupport":false},"linkedEditingRange":{"dynamicRegistration":true}},"window":{"showMessage":{"messageActionItem":{"additionalPropertiesSupport":true}},"showDocument":{"support":true},"workDoneProgress":true},"general":{"regularExpressions":{"engine":"ECMAScript","version":"ES2020"},"markdown":{"parser":"marked","version":"1.1.0"}}},"trace":"off","workspaceFolders":[{"uri":"file:///c%3A/src/myapp/app","name":"app"}]}}
[0m[36m[DEBUG] [0m> {"jsonrpc":"2.0","id":0,"result":{"capabilities":{"codeActionProvider":true,"completionProvider":{"resolveProvider":true,"triggerCharacters":["(","\n",",","@"]},"definitionProvider":true,"hoverProvider":true,"referencesProvider":true,"textDocumentSync":1}}}
[0m[36m[DEBUG] [0m< {"jsonrpc":"2.0","method":"initialized","params":{}}
[0m[36m[DEBUG] [0mJSON-RPC handshake completed
[0m[36m[DEBUG] [0mRunning language server with config root "\\\\?\\C:\\src\\myapp\\app\\src"
[0m[36m[DEBUG] [0mCreating lsp_state...
[0m[36m[DEBUG] [0mCreating lsp_state created!
[0m[36m[DEBUG] [0m< {"jsonrpc":"2.0","method":"textDocument/didOpen","params":{"textDocument":{"uri":"file:///c%3A/src/myapp/app/src/pages/Test.tsx","languageId":"typescriptreact","version":1,"text":"import { graphql } from \"react-relay\"\r\n\r\nconst query = graphql`\r\n    query Test_Query {\r\n        viewer {\r\n            user {\r\n                id\r\n            }\r\n        }\r\n    }\r\n`\r\n\r\nexport const Test = () => {\r\n    return null;\r\n}"}}}
[0m[36m[DEBUG] [0mProcessing task "InboundMessage(Notification(Notification { method: \"textDocument/didOpen\", params: Object({\"textDocument\": Object({\"languageId\": String(\"typescriptreact\"), \"text\": String(\"import { graphql } from \\\"react-relay\\\"\\r\\n\\r\\nconst query = graphql`\\r\\n    query Test_Query {\\r\\n        viewer {\\r\\n            user {\\r\\n                id\\r\\n            }\\r\\n        }\\r\\n    }\\r\\n`\\r\\n\\r\\nexport const Test = () => {\\r\\n    return null;\\r\\n}\"), \"uri\": String(\"file:///c%3A/src/myapp/app/src/pages/Test.tsx\"), \"version\": Number(1)})}) }))"
[0m[36m[DEBUG] [0mnotification received Notification { method: "textDocument/didOpen", params: Object({"textDocument": Object({"languageId": String("typescriptreact"), "text": String("import { graphql } from \"react-relay\"\r\n\r\nconst query = graphql`\r\n    query Test_Query {\r\n        viewer {\r\n            user {\r\n                id\r\n            }\r\n        }\r\n    }\r\n`\r\n\r\nexport const Test = () => {\r\n    return null;\r\n}"), "uri": String("file:///c%3A/src/myapp/app/src/pages/Test.tsx"), "version": Number(1)})}) }
[0m[36m[DEBUG] [0mlsp_method: textDocument/didOpen
[0m[36m[DEBUG] [0mlsp_type: notification
[0m[36m[DEBUG] [0mlsp_outcome: success
[0m[36m[DEBUG] [0m   0ms lsp_message_processing_time
[0m[36m[DEBUG] [0mtask InboundMessage(Notification(Notification { method: "textDocument/didOpen", params: Object({"textDocument": Object({"languageId": String("typescriptreact"), "text": String("import { graphql } from \"react-relay\"\r\n\r\nconst query = graphql`\r\n    query Test_Query {\r\n        viewer {\r\n            user {\r\n                id\r\n            }\r\n        }\r\n    }\r\n`\r\n\r\nexport const Test = () => {\r\n    return null;\r\n}"), "uri": String("file:///c%3A/src/myapp/app/src/pages/Test.tsx"), "version": Number(1)})}) })) completed in 0ms
[0m[36m[DEBUG] [0m< {"jsonrpc":"2.0","id":1,"method":"textDocument/codeAction","params":{"textDocument":{"uri":"file:///c%3A/src/myapp/app/src/pages/Test.tsx"},"range":{"start":{"line":0,"character":0},"end":{"line":0,"character":0}},"context":{"diagnostics":[]}}}
[0m[36m[DEBUG] [0mProcessing task "InboundMessage(Request(Request { id: RequestId(I32(1)), method: \"textDocument/codeAction\", params: Object({\"context\": Object({\"diagnostics\": Array([])}), \"range\": Object({\"end\": Object({\"character\": Number(0), \"line\": Number(0)}), \"start\": Object({\"character\": Number(0), \"line\": Number(0)})}), \"textDocument\": Object({\"uri\": String(\"file:///c%3A/src/myapp/app/src/pages/Test.tsx\")})}) }))"
[0m[36m[DEBUG] [0mrequest received Request { id: RequestId(I32(1)), method: "textDocument/codeAction", params: Object({"context": Object({"diagnostics": Array([])}), "range": Object({"end": Object({"character": Number(0), "line": Number(0)}), "start": Object({"character": Number(0), "line": Number(0)})}), "textDocument": Object({"uri": String("file:///c%3A/src/myapp/app/src/pages/Test.tsx")})}) }
[0m[36m[DEBUG] [0mlsp_method: textDocument/codeAction
[0m[36m[DEBUG] [0mlsp_type: request
[0m[36m[DEBUG] [0mlsp_outcome: error
[0m[36m[DEBUG] [0mlsp_error_message: Request Canceled
```

Is this a know issue or did I setup something incorrect? How can I provide more info?

Would love to see the extension go stable as soon as possible. It will surely be a massive improvement to our workflow. Thank you @tbezman for championing the extension and everyone else involved in the LSP development :)

### PR DESCRIPTION
**Pull request by tobias-tengler**
For every language server endpoint dealing with documents the code previously checked, whether the incoming file URI was part of the root directory:

```rust
uri.path().starts_with(state.root_dir().to_string_lossy().as_ref())
```

This works fine on Unix, since paths and URIs are "the same":

Root directory: `/home/user/my-app/`
File URI: `/home/user/my-app/subdir/file.ts`

For Windows you can see why the simple string `startsWith` might not suffice:

Root directory: `C:\user\my-app\`
File URI: `/c/user/my-app/subdir/file.ts`

In order to address this we now first convert the URI to a OS specific file path before doing the `startsWith`.

Furthermore [dunce](https://docs.rs/dunce/latest/dunce/) is being used to canonicalize paths, preventing paths to be converted to their UNC variant.

Closes #3918

### CODE REVIEW

File: compiler/crates/relay-lsp/src/server/lsp_request_dispatch.rs [hunk 1]
```diff
@@ -79,6 +80,14 @@ fn convert_to_lsp_response(
...
         },
         Err(runtime_error) => {
+            if let LSPRuntimeError::ExpectedError = runtime_error {
+                return ServerResponse {
+                    id,
+                    result: Some(Value::Null),
+                    error: None,
+                };
+            }
+
```
Comment (alunyov) [inline compiler/crates/relay-lsp/src/server/lsp_request_dispatch.rs:90]: Hmm, why do we need to change this? Is this related to the path?

**Review by alunyov — CHANGES_REQUESTED** [review]
Thank you!

The changes looks good. I have a small question about `convert_to_lsp_response` changes, how are they related to other changes in this PR?

File: compiler/crates/relay-lsp/src/server/lsp_request_dispatch.rs (hunk 1 above)
Comment (tobias-tengler) [reply compiler/crates/relay-lsp/src/server/lsp_request_dispatch.rs:90]: Wupps, just mentioned that in the original issue, but not the PR description 😅

If the file is not part of the directory the LSP operates on (as checked by the various ifs I made changes to), the methods return a `LSPRuntimeError::ExpectedError`. The `LSPRuntimeError` is converted to a `ResponseError` here: https://github.com/facebook/relay/blob/main/compiler/crates/relay-lsp/src/server/lsp_request_dispatch.rs#L82

In this conversion a `ExpectedError` will be mapped to `None`: https://github.com/facebook/relay/blob/main/compiler/crates/relay-lsp/src/lsp_runtime_error.rs#L22

The `None` is mapped to a `Request cancelled` error as a fallback I supose: https://github.com/facebook/relay/blob/main/compiler/crates/relay-lsp/src/server/lsp_request_dispatch.rs#L83-L87

The `Request cancelled` error sent me on a totally wrong track to tackle the underlying issue at first, since I thought the request to the LSP was being aborted only on Windows... So I opted to not return an error in the case of an expected error. I didn't find anything about how to handle expected errors in an LSP, so I just opted to return a `null` result. I wasn't sure how to handle it better as I'm a novice both regarding LSPs and Rust. Maybe one of you guys has a better idea how to surface expected errors correctly.

Comment (tobias-tengler) [reply compiler/crates/relay-lsp/src/server/lsp_request_dispatch.rs]: Done :)
